### DISCUSSION TITLE
Manually modify GITHUB_RUN_NUMBER for workflow

### DISCUSSION THREAD

**Discussion by Danielku15** [community/community#26709] [Actions]
<p>The GITHUB_RUN_NUMBER is currently a nice way of having a build number on your builds. I’m using it to build up a semver like <code>1.2.3-alpha.{number}</code>. Unfortunately this number resets if you change the workflow name and in some other scenarios. </p>
<p>This can lead to an issue if you deploy your package to registries like <a href="http://npmjs.com" rel="nofollow noopener">npmjs.com</a>. It would be great if we can manually set the new value of GITHUB_RUN_NUMBER within the repository. I could bump it again from 1 to the number I had before renaming my workflow file. </p>
<p>Another use case is if I bump my version to 1.2.4 I want to start again with <code>alpha.1</code>. Many CI servers expose this build number/counter and it can be reset manually if needed. GitHub actions seems to be missing this feature so far.</p>
<p>Is there maybe a workaround for this? I would like to avoid spoiling my Git history with constant version bumps from the CI environment just to update the build version within my package.json. </p>

**Comment by Danielku15** [comment]
<p>Hi <a class="mention" href="/mscoutermarsh">@mscoutermarsh</a> </p>
<p>Thanks for sharing this insight. GitHub actions still has a huge potential to grow and I have no doubts that sooner or later these scenarios will be covered by some action or via GitHub actions directly. </p>
<p>I thought also of adopting <a href="https://github.com/einaregilsson/build-number/blob/master/main.js" rel="nofollow noopener">this early action</a> to my needs. It existed before the official build number and uses git refs/tags to store the build counters. </p>

**Comment by tmorone-rezi** [comment]
<p>It would be very useful if we could run a one-time operation to set the current run number for the repo. I am migrating our CI/CD pipeline from bitrise. Since we published releases using the bitrise build number, it becomes troublesome to go “back in time” and using a build number that was already used. I imagine others will have similar problems as they migrate from another system.</p>

**Reply by hmartinez82** [reply]
Yes. I'm running this exactly problem (coming from Jenkins) in several repos :(

**Comment by hopeman15** [comment]
<p>We are currently migrating from Travis, this feature request would be a wonderful thing :+1:</p>

**Comment by mattgodbolt** [comment]
<p>This would be an amazing feature to have. On other build systems (e.g. TeamCity) we can edit the “next run number” which lets us do this kind of thing. Having the build number that GH displays in the actions view match up with the build number we put in our binaries would be so handy.</p>

**Comment by mlilback** [comment]
<p>I just published an action that does what most commenters seem to want. It lets you add a base number to the GITHUB_RUN_NUMBER. I use it to tag docker images and can’t have the number reset. So I just have to update the base number in the workflow when I edit the workflow.</p>
<p><a href="https://github.com/mlilback/build-number/releases/tag/v1" rel="noopener nofollow ugc">mlilback/build-number action</a></p>

**Reply by Muheli** [reply]
Thanks for sharing this clever solution.

**Comment by alonbl** [comment]
I was too very surprised that the `github.run_number` depends on the workflow name.

**Reply by filipopo** [reply]
necrobump but if you don't want that, there is a `github.run_id`

**Comment by rajbopched11** [comment]
@mscoutermarsh is this issue resolved yet?
It's 2023 and we are still awaiting this feature ¯\_(ツ)_/¯

**Reply by mscoutermarsh** [reply]
I don't work for GitHub anymore. Try emailing support.

**Reply by rajbopched11** [reply]
Meanwhile, a work-around which serves us well 🛠️ 🥇 

```yml
steps:
      - name: Set Build Number Environment Variable
        run: |
          echo "BUILD_NUMBER=$((${{ github.run_number }}+${{ github.run_attempt }}))" >> $GITHUB_ENV
          echo $BUILD_NUMBER
```

**Reply by filipopo** [reply]
> Meanwhile, a work-around which serves us well 🛠️ 🥇

Necrobump but wouldn't this start counting the builds from 2 instead of 1, you should probably have a -1 in there

**Comment by AlexNolasco** [comment]
perhaps by adding two numbers?
https://github.com/orgs/community/discussions/25386

Example

```   steps:
      - name: Increase build
        run: |
          echo "GITHUB_RUN_NUMBER_WITH_OFFSET=$(($GITHUB_RUN_NUMBER+1971))" >> $GITHUB_ENV
```

**Reply by silashansen** [reply]
It's a clever workround and definitely useful, but I am still missing a formal way of doing this, as this is such a common thing in other CI solutions.

**Reply by hmartinez82** [reply]
This is what most people are doing. But it sucks because I want to change the `run-name` based on this calculated values. So none of the two things that Jenkins has supported for ages, changing the next build number and allowing to change the `run-name` after the  build starts,  are not supported and I get in this deadlock!

**Comment by Amerousful** [comment]
Hello!
Is there any update in 2024? I faced the same problem during migration from another CI/CD tool. In the current system, our build/run id is a crucial thing and we have to maintain the same increment order instead of starting anew.

**Comment by Vacxe** [comment]
We using auto incremented **build number** on CI for each build. I think its very common approach on many platforms. As well managing `run_number` is common functionality for any build system. Could you please provide it for GHA in 2024 as well.

**Comment by abhilashindulkar** [comment]
Would love to see this feature implemented soon. This will help a lot. I can't rename/duplicate workflows every time I mess up with the build.

**Comment by dinith-rusiru** [comment]
Response#1

**Comment by rusik** [comment]
Just faced same issue as other commenters in here. Please add this feature asap 🙏

**Comment by Danielku15** [comment]
Considering all my findings on GHA over the years, nowadays I would phrase my feature request different. While GITHUB_RUN_NUMBER modifications would maybe solve one problem, they don't solve others. 

Not sure if its worth opening an on request in favor of this one but just to place it somewhere: 

The ability to have an custom persistent key-value store as part of GitHub Actions would solve many use-cases. There is currently now way workflows can store custom variables/values in a deterministic and persistent way. Additionally some new expressions to handle arithmetic operations and type conversions would be needed.

Assuming that such a feature is available, you could simply create your own personal GITHUB_RUN_NUMBER by storing a custom counter in this key-value store. 

Theoretically such a store exists with the new "Repository Variables" feature. But you need to use the REST API to get/set them and authenticate with proper permissions. It should be possible to create a custom action incrementing a custom counter as a step in your build scripts. You can prefix it with workflow or job names as desired. Modifying/Resetting values manually can be done via the GitHub UI. 

If the feature is very important for anyone, you should be able to create such an easily for your repository. e.g. use [github-script](https://github.com/actions/github-script) to get, increment and store a custom counter. You might also make it available under a environment variable or output for further steps to use.  https://docs.github.com/en/rest/actions/variables?apiVersion=2022-11-28

**Reply by piotrekkr** [reply]
You could try using repository level properties. Kind of hacky but may work as a storage. Problem may be concurrency but if workflow runs are not frequent or in parallel, then it should work okay I think.

You will probably need a fine grained PAT (or GH App token) with `"Custom properties" repository permissions (write)`.

Here is [api endpoint](https://docs.github.com/en/rest/repos/custom-properties?apiVersion=2022-11-28#create-or-update-custom-property-values-for-a-repository) you could use to update property for repo.

And here is docs about [custom repo properties](https://docs.github.com/en/organizations/managing-organization-settings/managing-custom-properties-for-repositories-in-your-organization)

**Comment by TimGustafson** [comment]
Hi!  This is also a problem for us.  We use the CI/CD's build number for our version code for Google Play apps (and others), and we've recently migrated from self-hosted Jenkins to GHA, but we're realizing now that there's no way for us to set the "next build number", which means we have to resort to something hacky as mentioned in this thread.

**Reply by Danielku15** [reply]
Theoretically you could use action variables for this purpose. They seem to be a proper storage nowadays for such needs of custom build variables. 

The variables are in [the `${{ vars.XXX }}` context](https://docs.github.com/en/actions/how-tos/write-workflows/choose-what-workflows-do/use-variables#using-the-vars-context-to-access-configuration-variable-values). Updating the variable requires [API usage](https://docs.github.com/en/rest/actions/variables?apiVersion=2022-11-28#update-a-repository-variable) though. But it's fairly easy via [`github-script`](https://github.com/actions/github-script). 

I guess it would be a matter of a few minutes to stitch together some custom node action which loads and persists build numbers with custom patterns using octokit (e.g. variables per branch or globally depending on your needs). You can look at the actions I linked below to get an idea. You can also store this action in a shared repository or put it into your repo directly. No need to publish it to the market place.       

With such a solution, at any time, you can go into the UI and modify the values to your needs (assuming you have the necessary permissions). 

Or if you are open to adapt your general strategy you could switch to alternative systems like: 

* [GitVersion](https://gitversion.net/docs/) is a widely used system on how to versionize your projects with a variety of automatic version numbering mechanisms. 
* https://github.com/marketplace/actions/build-tag-number Generates a sequential number using git tags as storage. 
* https://github.com/marketplace/actions/build-number-generator-action or https://github.com/marketplace/actions/build-number-manager commits a JSON file with the build numbers. 
* https://github.com/marketplace/actions/repo-build-number stores one single build number in a variable. 
* Use a timestamp suffix instead of a numeric one. 

GitVersion seem to be the most proper mechanism nowadays, but other solutions like using github variables or a file in the git repository also are quite proper and not necessarily "hacky". Using Git Tags is quite hacky though. 

Hope that helps finding a solution.

**Reply by hmartinez82** [reply]
`Theoretically you could use action variables for this purpose.` and `Updating the variable requires API usage though. But it's fairly easy via [github-script](https://github.com/actions/github-script).`

Yes, that's exactly what I had to do as a workaround. Of course we lost atomicity, for instance, if two runs of the same workflow are started at the same time. So in my opinion there's no way this is a solution, it's just a temporary hack.

**Reply by hmartinez82** [reply]
Oh, I just saw that [piotrekkr](https://github.com/piotrekkr) properly described everything above and why it's so hacky.

**Comment by filipopo** [comment]
+2 but I ended up going with this

```yaml
      - name: Compute semantic version
        id: version
        run: |
          BUILD_NUMBER=$(($GITHUB_RUN_NUMBER + $GITHUB_RUN_ATTEMPT - 2))
          echo "v=v1.$(($BUILD_NUMBER / 10)).$(($BUILD_NUMBER % 10))" >> $GITHUB_OUTPUT

      - name: Get version later
        run: echo ${{ steps.version.outputs.v }}
```

**Comment by SneWs** [comment]
Same here, this is a big problem when moving to GH from another solution.
